Kintsugi 🍵

paulhauner opened the pull request:
## Description

This is the canonical branch for the Kintsugi testnet.

**For instructions on running the Kintsugi testnet, see https://github.com/sigp/lighthouse/issues/2830.**

This PR contains the changes required for Lighthouse to run "the merge" specifications. It does not deliver a production-ready implementation of "the merge" for Lighthouse, instead it provides the necessary components for Lighthouse to run on testnets and progress towards a production implementation.

Where this PR touches code that will run on the pre-merge (i.e., current) Beacon Chain, any changes should be production ready. For post-merge components (e.g., `execution_layer`), some degree of unfinished-ness is acceptable. After all, "the merge" specification is not yet complete.

## TODO

*This section lists tasks required before this branch can be merged into `unstable`*.

N/A

ethDreamer committed:
Initial merge changes

Added Execution Payload from Rayonism Fork

Updated new Containers to match Merge Spec

Updated BeaconBlockBody for Merge Spec

Completed updating BeaconState and BeaconBlockBody

Modified ExecutionPayload<T> to use Transaction<T>

Mostly Finished Changes for beacon-chain.md

Added some things for fork-choice.md

Update to match new fork-choice.md/fork.md changes

ran cargo fmt

Added Missing Pieces in eth2_libp2p for Merge

fix ef test

Various Changes to Conform Closer to Merge Spec

michaelsproul committed:
Fix consensus, SSZ, tree hash & run merge EF tests (#2622)

* Update to v1.1.0-beta.4 (squash of #2548)

* SSZ, cached tree hash, EF tests

paulhauner committed:
Fix clippy lints on merge-f2f (#2626)

* Remove unchecked arith from ssz_derive

* Address clippy lints in block_verfication

* Use safe math for is_valid_gas_limit

michaelsproul committed:
Update merge consensus to v1.1.0-beta.5 (#2630)

paulhauner committed:
Handle merge fork in web3signer (#2631)

michaelsproul committed:
Update test vectors to v1.1.0 (#2642)

realbigsean committed:
Store execution block hash in fork choice (#2643)

* - Update the fork choice `ProtoNode` to include `is_merge_complete`
- Add database migration for the persisted fork choice

* update tests

* Small cleanup

* lints

* store execution block hash in fork choice rather than bool

ethDreamer committed:
Fork boundary fix (#2646)

* Fixed Gossip Topics on Fork Boundary

ethDreamer committed:
Finished Gossip Block Validation Conditions (#2640)

* Gossip Block Validation is Much More Efficient

Co-authored-by: realbigsean <seananderson33@gmail.com>

paulhauner committed:
[Merge] Implement `execution_layer` (#2635)

* Checkout serde_utils from rayonism

* Make eth1::http functions pub

* Add bones of execution_layer

* Modify decoding

* Expose Transaction, cargo fmt

* Add executePayload

* Add all minimal spec endpoints

* Start adding json rpc wrapper

* Finish custom JSON response handler

* Switch to new rpc sending method

* Add first test

* Fix camelCase

* Finish adding tests

* Begin threading execution layer into BeaconChain

* Fix clippy lints

* Fix clippy lints

* Thread execution layer into ClientBuilder

* Add CLI flags

* Add block processing methods to ExecutionLayer

* Add block_on to execution_layer

* Integrate execute_payload

* Add extra_data field

* Begin implementing payload handle

* Send consensus valid/invalid messages

* Fix minor type in task_executor

* Call forkchoiceUpdated

* Add search for TTD block

* Thread TTD into execution layer

* Allow producing block with execution payload

* Add LRU cache for execution blocks

* Remove duplicate 0x on ssz_types serialization

* Add tests for block getter methods

* Add basic block generator impl

* Add is_valid_terminal_block to EL

* Verify merge block in block_verification

* Partially implement --terminal-block-hash-override

* Add terminal_block_hash to ChainSpec

* Remove Option from terminal_block_hash in EL

* Revert merge changes to consensus/fork_choice

* Remove commented-out code

* Add bones for handling RPC methods on test server

* Add first ExecutionLayer tests

* Add testing for finding terminal block

* Prevent infinite loops

* Add insert_merge_block to block gen

* Add block gen test for pos blocks

* Start adding payloads to block gen

* Fix clippy lints

* Add execution payload to block gen

* Add execute_payload to block_gen

* Refactor block gen

* Add all routes to mock server

* Use Uint256 for base_fee_per_gas

* Add working execution chain build

* Remove unused var

* Revert "Use Uint256 for base_fee_per_gas"

This reverts commit 6c88f19ac45db834dd4dbf7a3c6e7242c1c0f735.

* Fix base_fee_for_gas Uint256

* Update execute payload handle

* Improve testing, fix bugs

* Fix default fee-recipient

* Fix fee-recipient address (again)

* Add check for terminal block, add comments, tidy

* Apply suggestions from code review

Co-authored-by: realbigsean <seananderson33@GMAIL.com>

* Fix is_none on handle Drop

* Remove commented-out tests

Co-authored-by: realbigsean <seananderson33@GMAIL.com>

paulhauner committed:
[Merge] Add serde impls for `Transactions` type (#2649)

* Start implemented serde for transactions

* Revise serde impl

* Add tests for transaction decoding

paulhauner committed:
[Merge] Add execution API test vectors from Geth (#2651)

* Add geth request vectors

* Add geth response vectors

* Fix clippy lints

realbigsean committed:
add automated docker build for merge-f2f branch (#2654)

pawanjay176 committed:
Add merge fork_epoch and fork_version to Config (#2663)

paulhauner committed:
Disable autotests for beacon_chain (#2658)

paulhauner committed:
Add `BeaconChainHarness` tests for The Merge (#2661)

* Start adding merge tests

* Expose MockExecutionLayer

* Add mock_execution_layer to BeaconChainHarness

* Progress with merge test

* Return more detailed errors with gas limit issues

* Use a better gas limit in block gen

* Ensure TTD is met in block gen

* Fix basic_merge tests

* Start geth testing

* Fix conflicts after rebase

* Remove geth tests

* Improve merge test

* Address clippy lints

* Make pow block gen a pure function

* Add working new test, breaking existing test

* Fix test names

* Add should_panic

* Don't run merge tests in debug

* Detect a tokio runtime when starting MockServer

* Fix clippy lint, include merge tests

paulhauner committed:
Misc changes for merge testnets (#2667)

* Thread eth1_block_hash into interop genesis state

* Add merge-fork-epoch flag

* Build LH with minimal spec by default

* Add verbose logs to execution_layer

* Add --http-allow-sync-stalled flag

* Update lcli new-testnet to create genesis state

* Fix http test

* Fix compile errors in tests

ethDreamer committed:
Fixed bugs for m3 readiness (#2669)

* Fixed bugs for m3 readiness

* woops

* cargo fmt..

paulhauner committed:
Update lcli pubkey replace command (#2677)

paulhauner committed:
Disable notifier logging from dummy eth1 backend (#2680)

paulhauner committed:
[Merge] Block validator duties when EL is not ready (#2672)

* Reject some HTTP endpoints when EL is not ready

* Restrict more endpoints

* Add watchdog task

* Change scheduling

* Update to new schedule

* Add "syncing" concept

* Remove RequireSynced

* Add is_merge_complete to head_info

* Cache latest_head in Engines

* Call consensus_forkchoiceUpdate on startup

paulhauner committed:
[Merge] Optimistic EL verification (#2683)

* Ignore payload errors

* Only return payload handle on valid response

* Push some engine logs down to debug

* Push ee fork choice log to debug

* Push engine call failure to debug

* Push some more errors to debug

* Fix panic at startup

pawanjay176 committed:
Fix db paths when datadir is relative (#2682)

paulhauner committed:
[Merge] Optimistic Sync: Stage 1 (#2686)

* Add payload verification status to fork choice

* Pass payload verification status to import_block

* Add valid back-propagation

* Add head safety status latch to API

* Remove ExecutionLayerStatus

* Add execution info to client notifier

* Update notifier logs

* Change use of "hash" to refer to beacon block

* Shutdown on invalid finalized block

* Tidy, add comments

* Fix failing FC tests

* Allow blocks with unsafe head

* Fix forkchoiceUpdate call on startup

realbigsean committed:
v1.1.1 spec updates (#2684)

* update initializing from eth1 for merge genesis

* read execution payload header from file lcli

* add `create-payload-header` command to `lcli`

* fix base fee parsing

* Apply suggestions from code review

* default `execution_payload_header` bool to false when deserializing `meta.yml` in EF tests

Co-authored-by: Paul Hauner <paul@paulhauner.com>

paulhauner committed:
Accept TTD override as decimal (#2676)

paulhauner committed:
Kintsugi rebase patches (#2769)

* Freshen Cargo.lock

* Fix gossip worker

* Update map_fork_name_with

paulhauner committed:
Move merge-f2f docker to kintsugi (#2774)

realbigsean commented:
> Upgrade to v1.1.4 spec tests, fixing any consensus changes.

I can work on  this

realbigsean committed:
Fix arbitrary check kintsugi (#2777)

pawanjay176 committed:
Increase network limits (#2796)

Fix max packet sizes

Fix max_payload_size function

Add merge block test

Fix max size calculation; fix up test

Clear comments

Add a payload_size_function

Use safe arith for payload calculation

Return an error if block too big in block production

Separate test to check if block is over limit

realbigsean committed:
1.1.5 merge spec tests (#2781)

* Fix arbitrary check kintsugi

* Add merge chain spec fields, and a function to determine which constant to use based on the state variant

* increment spec test version

* Remove `Transaction` enum wrapper

* Remove Transaction new-type

* Remove gas validations

* Add `--terminal-block-hash-epoch-override` flag

* Increment spec tests version to 1.1.5

* Remove extraneous gossip verification https://github.com/ethereum/consensus-specs/pull/2687

* - Remove unused Error variants
- Require both "terminal-block-hash-epoch-override" and "terminal-block-hash-override" when either flag is used

* - Remove a couple more unused Error variants

Co-authored-by: Paul Hauner <paul@paulhauner.com>

pawanjay176 committed:
Fix Uint256 deserialization (#2786)

* Change base_fee_per_gas to Uint256

* Add custom (de)serialization to ExecutionPayload

* Fix errors

* Add a quoted_u256 module

* Remove unused function

* lint

* Add test

* Remove extra line

Co-authored-by: Paul Hauner <paul@paulhauner.com>

paulhauner committed:
Remove old uses of testnet

paulhauner commented:
I just rebased this branch onto `unstable`. It will fail CI until I push another fix (https://github.com/sigp/lighthouse/pull/2799).

michaelsproul commented on beacon_node/execution_layer/src/engine_api/http.rs:
```diff
@@ -0,0 +1,1061 @@
+                .is_err()
+        );
+
+        /*
+         * Check for transaction too large
+         */
+
+        use eth2_serde_utils::hex;
+
+        let num_max_bytes = <MainnetEthSpec as EthSpec>::MaxBytesPerTransaction::to_usize();
+        let max_bytes = (0..num_max_bytes).map(|_| 0_u8).collect::<Vec<_>>();
+        let too_many_bytes = (0..=num_max_bytes).map(|_| 0_u8).collect::<Vec<_>>();
+        decode_transactions::<MainnetEthSpec>(
+            serde_json::to_value(&[hex::encode(&max_bytes)]).unwrap(),
+        )
+        .unwrap();
+        assert!(decode_transactions::<MainnetEthSpec>(
+            serde_json::to_value(&[hex::encode(&too_many_bytes)]).unwrap()
+        )
+        .is_err());
```
This test is the cause of the CI failure on Windows. It allocates 2GB of memory (as 1GB + 1GB) due to `MaxBytesPerTransaction` now being equal to 1GB.

The `STATUS_STACK_BUFFER_OVERRUN` error code identified by Paul [here](https://github.com/sigp/lighthouse/pull/2781#issuecomment-964787275) is indicative of a "fast abort" on Windows, and indeed that stack trace also includes a line: `memory allocation of 2147483650 bytes failed`.

We could try testing the two 1GB transactions separately (with `drop` in between), but even then 1GB is still a lot to allocate and it might still fail. I'll raise a PR to try it I guess...

More on the `STATUS_STACK_BUFFER_OVERRUN`: https://devblogs.microsoft.com/oldnewthing/20190108-00/?p=100655

michaelsproul replied:
I think the lowest we can go is 3GB allocated, because the hex encoding doubles the size of the transaction (1GB bytes + 2GB hex bytes). The CI machines allegedly have 7GB of RAM so this might be OK: https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources

michaelsproul replied:
Can probably do 2GB with some string hacks. Trying that now

paulhauner replied:
Resolved via #2809 :pray:

paulhauner committed:
Fixes after rebasing Kintsugi onto unstable (#2799)

* Fix fork choice after rebase

* Remove paulhauner warp dep

* Fix fork choice test compile errors

* Assume fork choice payloads are valid

* Add comment

* Ignore new tests

* Fix error in test skipping

paulhauner committed:
Skip memory intensive engine test (#2809)

* Allocate less memory (3GB) in engine tests

* Run cargo format

* Remove tx too large test

Co-authored-by: Michael Sproul <michael@sigmaprime.io>

paulhauner committed:
Implement engine API v1.0.0-alpha.4 (#2810)

* Added ForkchoiceUpdatedV1 & GetPayloadV1

* Added ExecutePayloadV1

* Added new geth test vectors

* Separated Json Object/Serialization Code into file

* Deleted code/tests for Requests Removed from spec

* Finally fixed serialization of null '0x'

* Made Naming of JSON Structs Consistent

* Fix clippy lints

* Remove u64 payload id

* Remove unused serde impls

* Swap to [u8; 8] for payload id

* Tidy

* Adjust some block gen return vals

* Tidy

* Add fallback when payload id is unknown

* Remove comment

Co-authored-by: Mark Mackey <mark@sigmaprime.io>

paulhauner committed:
Ensure difficulty/hash/epoch overrides change the `ChainSpec` (#2798)

* Unify loading of eth2_network_config

* Apply overrides at lighthouse binary level

* Remove duplicate override values

* Add merge values to existing net configs

* Make override flags global

* Add merge fields to testing config

* Add one to TTD

* Fix failing engine tests

* Fix test compile error

* Remove TTD flags

* Move get_eth2_network_config

* Fix warn

* Address review comments

paulhauner committed:
Kintsugi on_merge_block tests (#2811)

* Start v1.1.5 updates

* Implement new payload creation logic

* Tidy, add comments

* Remove unused error enums

* Add validate payload for gossip

* Refactor validate_merge_block

* Split payload verification in per block processing

* Add execute_payload

* Tidy

* Tidy

* Start working on new fork choice tests

* Fix failing merge block test

* Skip block_lookup_failed test

* Fix failing terminal block test

* Fixes from self-review

* Address review comments

ethDreamer committed:
Removed PowBlock struct that never got used (#2813)

michaelsproul commented on Makefile:
```diff
@@ -23,9 +23,9 @@ FORKS=phase0 altair
 install:
 ifeq ($(PORTABLE), true)
-	cargo install --path lighthouse --force --locked --features portable
+	cargo install --path lighthouse --force --locked --features portable,spec-minimal
```
We can probably remove this now that the devnets are no longer using minimal specs, right? It's also fine if it stays, just means slightly more compile time for people compiling from source

paulhauner replied:
Agreed, I'll grab this in a cumulative PR.

If it turns out people want `spec-minimal` for interop we can always resurrect the `kintsugi` branch and just apply small changes like this.

paulhauner replied:
Resolved in #2831

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3173,6 +3259,30 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+                    .try_send(ShutdownReason::Failure(
+                        "Finalized block has an invalid execution payload.",
+                    ))
+                    .map_err(BeaconChainError::InvalidFinalizedPayloadShutdownError)?;
```
This isn't relevant to non-merge networks but I think we should `return` here so that this function doesn't continue running while the node shuts down

paulhauner replied:
Resolved in #2831

paulhauner commented:
I just rebased to fix a conflict with `unstable` and to fix a merge (instead of squash-merge) of #2796.

This will fail CI until we merge https://github.com/sigp/lighthouse/pull/2829.

michaelsproul commented on consensus/proto_array/src/proto_array_fork_choice.rs:
```diff
@@ -252,6 +295,22 @@ impl ProtoArrayForkChoice {
     }
 
+    /// Only used for SSZ deserialization of the persisted fork choice during the database migration
+    /// from schema 4 to schema 5.
```
```suggestion
    /// from schema 5 to schema 6.
```

michaelsproul commented on consensus/proto_array/src/ssz_container.rs:
```diff
@@ -17,6 +18,35 @@ pub struct SszContainer {
 }
 
+/// Only used for SSZ deserialization of the persisted fork choice during the database migration
+/// from schema 4 to schema 5.
```
```suggestion
/// from schema 5 to schema 6.
```

michaelsproul commented on consensus/types/Cargo.toml:
```diff
@@ -38,7 +38,7 @@ tempfile = "3.1.0"
 rusqlite = { version = "0.25.3", features = ["bundled"], optional = true }
 arbitrary = { version = "1.0", features = ["derive"], optional = true }
-eth2_serde_utils = "0.1.0"
+eth2_serde_utils = { path = "../serde_utils" }
```
This could/should be a top-level patch

michaelsproul reviewed (approved):
I have reviewed the `./consensus` changes thoroughly and I'm happy to merge this to `unstable`.

paulhauner commented on beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs:
```diff
@@ -559,6 +561,12 @@ fn handle_v2_response<T: EthSpec>(
                 ForkName::Base => Ok(Some(RPCResponse::BlocksByRange(Box::new(
                     SignedBeaconBlock::Base(SignedBeaconBlockBase::from_ssz_bytes(decoded_buffer)?),
                 )))),
+                // TODO: check this (though it seems okay)
```
Needs to be checked and removed.

divagant-martian replied:
looks ok to me

paulhauner replied:
Resolved via #2836

paulhauner commented on beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs:
```diff
@@ -569,6 +577,12 @@ fn handle_v2_response<T: EthSpec>(
                 ForkName::Base => Ok(Some(RPCResponse::BlocksByRoot(Box::new(
                     SignedBeaconBlock::Base(SignedBeaconBlockBase::from_ssz_bytes(decoded_buffer)?),
                 )))),
+                // TODO: check this (though it seems right)
```
Needs to be checked and removed.

divagant-martian replied:
looks ok too

paulhauner replied:
Resolved via #2836

paulhauner commented on beacon_node/src/config.rs:
```diff
@@ -36,20 +38,12 @@ pub fn get_config<E: EthSpec>(
-            fs::remove_dir_all(chain_db)
-                .map_err(|err| format!("Failed to remove chain_db: {}", err))?;
-        }
+        fs::remove_dir_all(client_config.get_db_path())
```
Should we continue to check if the path exists before removing it?

michaelsproul replied:
Yeah good catch, Mac added this in #2747

paulhauner commented on beacon_node/src/config.rs:
```diff
@@ -36,20 +38,12 @@ pub fn get_config<E: EthSpec>(
-            fs::remove_dir_all(freezer_db)
-                .map_err(|err| format!("Failed to remove chain_db: {}", err))?;
-        }
+        fs::remove_dir_all(client_config.get_freezer_db_path())
```
As above.

paulhauner committed:
Bump crate versions (#2829)

divagant-martian commented on beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs:
```diff
@@ -375,31 +375,33 @@ fn handle_error<T>(
+                    // NOTE: If you are adding another fork type here, be sure to modify the
+                    //       `fork_context.to_context_bytes()` function to support it as well!
+                    SignedBeaconBlock::Merge { .. } => {
+                        // TODO: check this
```
this looks fine to me reading the spec. Do we have doubts about this?

paulhauner replied:
It looks good to me too :)

paulhauner replied:
Resolved via #2836

divagant-martian commented on beacon_node/lighthouse_network/src/rpc/protocol.rs:
```diff
@@ -194,9 +206,10 @@ impl RpcLimits {
-    /// Returns true if the given length is out of bounds, false otherwise.
-    pub fn is_out_of_bounds(&self, length: usize) -> bool {
-        length > self.max || length < self.min
+    /// Returns true if the given length is is greater than `MAX_RPC_SIZE` or out of
```
```suggestion
    /// Returns true if the given length is greater than `MAX_RPC_SIZE` or out of
```

paulhauner replied:
Resolved via #2836

divagant-martian reviewed (approved):
I checked `./lighthouse_network` and `./network`. it looks good to me

paulhauner committed:
Kintsugi review comments (#2831)

* Fix makefile

* Return on invalid finalized block

* Fix todo in gossip scoring

* Require --merge for --fee-recipient

* Bump eth2_serde_utils

* Change schema versions

* Swap hash/uint256 test_random impls

* Use default for ExecutionPayload::empty

* Check for DBs before removing

* Remove kintsugi docker image

* Fix CLI default value

ethDreamer committed:
Cleanup Comments & Fix get_pow_block_hash_at_ttd() (#2835)

paulhauner committed:
Kintsugi Diva comments (#2836)

* Remove TODOs

* Fix typo

michaelsproul commented on common/slot_clock/src/lib.rs:
```diff
@@ -99,4 +102,14 @@ pub trait SlotClock: Send + Sync + Sized + Clone {
         self.slot_duration() * 2 / 3
     }
+
+    /// An implementation of the method described in the consensus spec here:
+    ///
+    /// https://github.com/ethereum/consensus-specs/blob/dev/specs/merge/beacon-chain.md#compute_timestamp_at_slot
+    fn compute_timestamp_at_slot(&self, slot: Slot) -> Option<u64> {
+        let slots_since_genesis = slot.as_u64().checked_sub(self.genesis_slot().as_u64())?;
+        slots_since_genesis
+            .checked_mul(self.slot_duration().as_secs())
+            .and_then(|since_genesis| self.genesis_duration().as_secs().checked_add(since_genesis))
+    }
```
I just realised this is the same as `SlotClock::start_of(slot: Slot)` but returning a `u64` rather than a `Duration`. I think we could simplify things a little by removing it and calling `start_of(slot).map(|d| d.as_secs())` instead

paulhauner committed:
Support legacy data directories (#2846)

paulhauner committed:
Remove duplicate slot_clock method (#2842)

pawanjay176 committed:
Restrict network limits based on merge fork epoch (#2839)

paulhauner commented:
I've just rebased to fix a merge conflict.
